squad-server/index.js

```diff
@@ -17,31 +17,135 @@ import Rcon from 'rcon/squad';
 import { SQUADJS_VERSION } from './utils/constants.js';
 import { SquadLayers } from './utils/squad-layers.js';
 
-import plugins from './plugins/index.js';
+import { ServerConfig } from '../serverconfig.js';
+import glob from 'glob';
 
 const __dirname = path.dirname(fileURLToPath(import.meta.url));
 
 export default class SquadServer extends EventEmitter {
-  constructor(options = {}) {
+  constructor() {
     super();
+    Logger.verbose('SquadServer', 1, 'Creating SquadServer...');
 
-    for (const option of ['host', 'queryPort'])
-      if (!(option in options)) throw new Error(`${option} must be specified.`);
+    this.layerHistory = [];
+    this.players = [];
+    this.plugins = [];
+    this.connectors = {};
+    const config = ServerConfig.getInstance().config;
+
+    Object.entries(config.verboseness).forEach(([module, verboseness]) => Logger.setVerboseness(module, verboseness));
+
+    // ?? SquadLayers is present both in SquadServer and connectors array
+    this.squadLayers = new SquadLayers(config.server.squadLayersSource);
+    this.rcon = this.setupRCON(config.server);
+    this.logParser = this.setupLogParser(config.server);
+    this.setupUpdateIntervals();
+
+    // move squadLayers pull to the SquadLayers constructor ?
+    this.squadLayers.pull();
+    this.setupPlugins(config.plugins).then(pluginInstances => {
+      this.plugins = pluginInstances;
+      
+      Logger.verbose('SquadServer', 1, `Number of Plugins online [${this.plugins.length}]`);
+      this.plugins.forEach(plugin => {
+        Logger.verbose('SquadServer', 2, `Plugin online: ${plugin.constructor.name}`);
+      });
+    });
+  }
 
-    this.options = options;
+  async setupPlugins(pluginsConfig) {
+    Logger.verbose('SquadServer', 1, 'Seting up plugins...');
+    // Get all potential plugin files from plugin directory
+    return await glob.sync('./plugins/*.js', { cwd: './squad-server/'}).reduce(async (plugins, file) => {
+      // import potential plugin class from file name
+      const { default: plugin } = await import(file);
+      plugins = await plugins;
+
+      // check if BasePlugin is implemented to make shore its a plugin
+      if (plugin.hasOwnProperty('optionsSpecification')) {
+        // recover plugins configuration
+        const currentConf = pluginsConfig.find(singleConfig => singleConfig.plugin === plugin.name);
+
+        // if config is absent plugin is disabled or unconfigured
+        if (currentConf) {
+          const pluginOptions = await this.buildPluginOptions(plugin.optionsSpecification, currentConf);
+          var pluginInstance;
+          // try creating plugin instance
+          // if failed dont crash the aplication just log and proceed with next plugin
```

Comment:
No, the config should verify before even trying to initiate anything that all config is in order.
It should write an error, and not crash the application.
I dont see any app that crashing it would be a good idea in genera.

Reply:
In this case, no. You could consider a miss configuration the equivalent of a compilation error... we know it's going to cause issues so we should not allow the program to run without it being corrected. Of course, runtime errors should be handled without the program closing because we want it to recover, but these are usually unpredictable so cannot be raised beforehand/when the program is first run.

Ultimately, I don't want an error to go unnoticed resulting in a plugin not working. Not everyone using SquadJS is likely to check logs for warnings or hidden errors, especially when most users aren't knowledgable in programming and the like. 

Furthermore, if someone is frequently starting SquadJS's `watch()` method whilst debugging an invalid config then it's likely to cause issues, such as being locked out of Squad's RCON for reconnecting too many times or spamming in-game announcements due to plugins losing their state/cooldowns.